internals/src/compact_size.rs

```diff
@@ -0,0 +1,310 @@
+///
+/// # Returns
+///
+/// The `CompactSize` if one exists as well as the number of bytes used to encode it.
+///
+/// # Example
+///
+/// ```
+/// use bitcoin_internals::compact_size;
+///
+/// // Extra bytes are ignored.
+/// let (compact, _) = compact_size::decode(&[0x20, 0xAB, 0xCD]).unwrap();
+/// assert_eq!(compact, 32);
+/// ```
+pub fn decode(slice: &[u8]) -> Result<(u64, usize), DecodeError> {
+    if slice.is_empty() {
+        return Err(DecodeError::Empty);
+    }
+
+    let res = match slice[0] {
+        0xFF => {
+            const SIZE: usize = 9;
+            if slice.len() < SIZE {
+                return Err(TooShortError::new(slice, SIZE).into());
+            };
+
+            let mut bytes = [0_u8; SIZE - 1];
+            bytes.copy_from_slice(&slice[1..SIZE]);
+            (u64::from_le_bytes(bytes), SIZE)
+        }
+        0xFE => {
+            const SIZE: usize = 5;
+            if slice.len() < SIZE {
+                return Err(TooShortError::new(slice, SIZE).into());
+            };
+
+            let mut bytes = [0_u8; SIZE - 1];
+            bytes.copy_from_slice(&slice[1..SIZE]);
+            let v = u32::from_le_bytes(bytes);
+            (u64::from(v), SIZE)
+        }
+        0xFD => {
+            const SIZE: usize = 3;
+            if slice.len() < SIZE {
+                return Err(TooShortError::new(slice, SIZE).into());
+            };
+
+            let mut bytes = [0_u8; SIZE - 1];
+            bytes.copy_from_slice(&slice[1..SIZE]);
+            let v = u16::from_le_bytes(bytes);
+            (u64::from(v), SIZE)
+        }
+        n => (u64::from(n), 1),
+    };
+    Ok(res)
+}
+
+/// Error returned when decoding a `CompactSize`.
+///
+/// In other words the object being serialized is too big to go over the wire.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum DecodeError {
+    /// Tried to decode a `CompactSize` from an empty slice.
+    Empty,
+    /// Slice was too short to decode.
+    TooShort(TooShortError),
+}
+
+crate::impl_from_infallible!(DecodeError);
+
+impl fmt::Display for DecodeError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use DecodeError::*;
+
+        match *self {
+            Empty => write!(f, "tried to decode a CompactSize from an empty slice"),
+            TooShort(ref e) => write_err!(f, "decode"; e),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for DecodeError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        use DecodeError::*;
+
+        match *self {
+            Empty => None,
+            TooShort(ref e) => Some(e),
+        }
+    }
+}
+
+impl From<TooShortError> for DecodeError {
+    fn from(e: TooShortError) -> Self { Self::TooShort(e) }
+}
+
+/// Error returned if a slice is too short based on the initial byte.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct TooShortError {
+    initial_byte: u8,
+    expected_minimum_length: usize,
```

Comment:
These are redundant since you can compute one from the other.

Reply:
Who cares? There is no perf hit in having an extra field in an error type, right? Its more brain-dead-simple to carry the redundant info than having more byte -> length conversions.

Reply:
Technically, there actually is. Size of error types causes the whole `Result` to be larger if it's larger than the `Ok` variant and this slows down `Result` operations downstream. We generally don't pay too much attention to this since bad errors are worse than bad performance in most cases but if we can reasonably compress the information we should. The return type of the function is `u64` which makes the `Result` ~twice as large as it would be if it was just the value itself. However it'd be really strange if this simple function didn't get inlined which may remove the cost entirely.

And really, the information we need can fit just **4 bits** (see below). But I'm not asking you to implement that madness, I'm only asking to fit it into ~`u64`. :)

Side note: it annoys me that `ArrayVec` has to use `usize` for length tracking even if `CAP` is small enough to fit `u8`.

4 bits because the errors that can possibly exist are only these:

* The input is empty
* The input encodes two-byte requirement and one of the bytes is missing
* The input encodes four-byte requirement and at most four bytes are missing - four combinations
* The input encodes eight-byte requirement and at most eight bytes are missing - eight combinations

Which gives us 1 + 1 + 4 + 8 == 14 combination which can be encoded by 4 bits.

Yes, it was a fun puzzle to find the true lower bound and not just any solution that fits into a byte. :)

BTW the most efficient way you can write it in Rust is to enumerate all the possibilities as `enum` because that also gives niche optimizations.

Reply:
I didn't read Kix's crazy analysis, but definitely +1 to minimizing error type sizes where it's reasonably possible.

In `miniscript` we have a massive error types everywhere and it's a big problem for us. It's not so big a problem here but we shouldn't get complacent.

Reply:
Cool, thanks for the explanation. TIL.

Reply:
> In `miniscript` we have a massive error types everywhere and it's a big problem for us. It's not so big a problem here but we shouldn't get complacent.

TIL about the [result_large_err](https://rust-lang.github.io/rust-clippy/master/index.html#/result_larg) clippy lint, which fires on errors that are over 128 bytes, but which can be configured to fire at a lower threshold.